interfaces/prompting/constraints.go

```diff
@@ -97,17 +129,333 @@ func (c *Constraints) Match(path string) (bool, error) {
 	return match, nil
 }
 
-// ContainPermissions returns true if the constraints include every one of the
-// given permissions.
-func (c *Constraints) ContainPermissions(permissions []string) bool {
-	for _, perm := range permissions {
-		if !strutil.ListContains(c.Permissions, perm) {
+// ReplyConstraints hold information about the applicability of a reply to
+// particular paths or permissions. Upon receiving the reply, snapd converts
+// ReplyConstraints to Constraints.
+type ReplyConstraints struct {
+	PathPattern *patterns.PathPattern `json:"path-pattern"`
+	Permissions []string              `json:"permissions"`
```

Comment:
We don't check this, and we don't particularly care if the client response with something like `allow ["read", "read"]`. `ReplyConstraints` are validated during `ToConstraints`, where we iterate through the permissions in the `Permissions` list and add a permission entry to the permission map for each permission. Duplicate permissions in the list means we'll re-add the same `key:value` to the map twice, but the key and value are identical in this case, so that's no problem.